### PR TITLE
`Integrated code lifecycle`: Allow force pushes to template, solution and tests repositories for editors

### PR DESCRIPTION
**Pull request by laurenzfb**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.

#### Changes affecting Programming Exercises
- [x] **High priority**: I tested **all** changes and their related features with **all** corresponding user types on a test server configured with the **integrated lifecycle setup** (LocalVC and LocalCI).

### Motivation and Context

Instructors and editors should be able to force push to template, solution and tests repositories.

### Description

I adapted the check that prevents force pushes so that it allows force pushes if the user is at least editor in the course and the repository is either template, solution or tests. I also added a test that verifies the new functionality.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Programming exercise

Repeat these steps for template, solution and tests
1. Clone the repository locally in git
2. Open the repository in the Artemis code editor and submit changes
3. In the local repository, commit a new change and try to push (should not work as there are changes on remote)
4. Try to force push `git push --force` (should work)

For assignment repository:
1. Participate in the exercise
2. Clone the repository locally in git
3. Open the repository in the Artemis code editor and submit changes
4. In the local repository, commit a new change and try to push (should not work as there are changes on remote)
5. Try to force push `git push --force` (should not work)

#### Performance Review
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [ArtemisGitServlet.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6268/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.config.localvcci/ArtemisGitServlet.html) | 100% | ✅ |
| [LocalVCPostPushHook.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6268/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.localvc/LocalVCPostPushHook.html) | 75% | ✅ |
| [LocalVCPrePushHook.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6268/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.localvc/LocalVCPrePushHook.html) | 78% | ✅ |
| [LocalVCServletService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6268/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.localvc/LocalVCServletService.html) | 92% | ✅ |

## Summary by CodeRabbit

- **New Features**
	- Enhanced security by checking user permissions for force push operations.
- **Refactor**
	- Improved authentication and authorization logic for better readability and separation of concerns.
- **Tests**
	- Added and refactored integration tests to cover new scenarios, including force push attempts by different user roles.

### CODE REVIEW

File: src/test/java/de/tum/in/www1/artemis/localvcci/LocalVCIntegrationTest.java [hunk 1]
```diff
@@ -191,38 +196,83 @@ void testUserTriesToDeleteBranch() throws GitAPIException {
...
+            return pushResultForce.getRemoteUpdates().iterator().next();
+        }
+        finally {
+            if (secondLocalGit != null) {
+                secondLocalGit.close();
+            }
+            if (tempDirectory != null) {
+                FileUtils.deleteDirectory(tempDirectory.toFile());
+            }
+        }
```
[the remark this answers is not in the document]
Comment (laurenzfb) [reply src/test/java/de/tum/in/www1/artemis/localvcci/LocalVCIntegrationTest.java:269]: no

**Review by reschandreas — APPROVED** [review]
demonstrated in testing session works as expected, code also lgtm!

**Review by rstief — APPROVED** [review]
Code LGTM 👀

**Review by milljoniaer — APPROVED** [review]
```bash
# Demonstrated in testing session
approve --force
```

**Pull request merged**